internal/querycoordv2/meta/resource_manager.go

```diff
@@ -0,0 +1,593 @@
+	ErrRemoveResourceGroupFromStore = errors.New("failed to remove resource group from store")
+	ErrRecoverResourceGroupToStore  = errors.New("failed to recover resource group to store")
+	ErrNodeNotAssignToRG            = errors.New("node hasn't been assign to any resource group")
+	ErrRGNameIsEmpty                = errors.New("resource group name couldn't be empty")
+	ErrDeleteDefaultRG              = errors.New("delete default rg is not permitted")
+	ErrDeleteNonEmptyRG             = errors.New("delete non-empty rg is not permitted")
+	ErrNodeNotExist                 = errors.New("node does not exist")
+	ErrNodeStopped                  = errors.New("node has been stopped")
+)
+
+var DefaultResourceGroupName = "__default_resource_group"
+
+type ResourceGroup struct {
+	nodes    UniqueSet
+	capacity int
+}
+
+func NewResourceGroup(capacity int) *ResourceGroup {
+	rg := &ResourceGroup{
+		nodes:    typeutil.NewUniqueSet(),
+		capacity: capacity,
+	}
+
+	return rg
+}
+
+// assign node to resource group
+func (rg *ResourceGroup) assignNode(id int64) error {
+	if rg.containsNode(id) {
+		return ErrNodeAlreadyAssign
+	}
+
+	rg.nodes.Insert(id)
+	rg.capacity++
+
+	return nil
+}
+
+// unassign node from resource group
+func (rg *ResourceGroup) unassignNode(id int64) error {
+	if !rg.containsNode(id) {
+		// remove non exist node should be tolerable
+		return nil
+	}
+
+	rg.nodes.Remove(id)
+	rg.capacity--
+
+	return nil
+}
+
+func (rg *ResourceGroup) handleNodeUp(id int64) error {
+	if rg.LackOfNodes() == 0 {
+		return ErrRGIsFull
+	}
+
+	if rg.containsNode(id) {
+		return ErrNodeAlreadyAssign
+	}
+
+	rg.nodes.Insert(id)
+	return nil
+}
+
+func (rg *ResourceGroup) handleNodeDown(id int64) error {
+	if !rg.containsNode(id) {
+		// remove non exist node should be tolerable
+		return nil
+	}
+
+	rg.nodes.Remove(id)
+	return nil
+}
+
+func (rg *ResourceGroup) LackOfNodes() int {
+	return rg.capacity - len(rg.nodes)
+}
+
+func (rg *ResourceGroup) containsNode(id int64) bool {
+	return rg.nodes.Contain(id)
+}
+
+func (rg *ResourceGroup) GetNodes() []int64 {
+	return rg.nodes.Collect()
+}
+
+func (rg *ResourceGroup) GetCapacity() int {
+	return rg.capacity
+}
+
+type ResourceManager struct {
+	groups  map[string]*ResourceGroup
+	store   Store
+	nodeMgr *session.NodeManager
+
+	rwmutex sync.RWMutex
+}
+
+func NewResourceManager(store Store, nodeMgr *session.NodeManager) *ResourceManager {
+	groupMap := make(map[string]*ResourceGroup)
+	groupMap[DefaultResourceGroupName] = NewResourceGroup(1000000)
+	return &ResourceManager{
+		groups:  groupMap,
+		store:   store,
+		nodeMgr: nodeMgr,
+	}
+}
+
+func (rm *ResourceManager) AddResourceGroup(rgName string) error {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+	if len(rgName) == 0 {
+		return ErrRGNameIsEmpty
+	}
+
+	if rm.groups[rgName] != nil {
+		return ErrRGAlreadyExist
+	}
+
+	rm.groups[rgName] = NewResourceGroup(0)
+	err := rm.saveResourceGroupInStore(rgName)
```

Comment:
persist it first, store in memory later, better to switch the two lines 160, 161